### PR TITLE
iss18

### PR DESCRIPTION
**Pull request by xavierfav**
In order to prevent getting duplicates of votes from the validation form:

- Added disable of the button after submitting (client side)
- Added transaction and check if annotation_id is already in the database before saving votes (server side)

### CODE REVIEW

**Comment by alastair** [comment]
Please give this pull request a descriptive title, and in the comment say that it relates to issue #18 (so that github links the two).

Sometimes people put `[WIP]` at the beginning of the pull request title (Work in progress) so that others know not to review it yet.

File: datasets/views.py [hunk 1]
```diff
@@ -155,20 +156,25 @@ def contribute_validate_annotations_category(request, short_name, node_id):
...
-                    # Save votes for annotations
-                    Vote.objects.create(
-                        created_by=request.user,
-                        vote=float(form.cleaned_data['vote']),
-                        visited_sound=form.cleaned_data['visited_sound'],
-                        annotation_id=form.cleaned_data['annotation_id'],
-                    )
+                    annotation_id = form.cleaned_data['annotation_id']
+                    check = Vote.objects.filter(created_by=request.user,
+                                               annotation_id=annotation_id)
```
Comment (alastair) [inline datasets/views.py]: this is not aligned with the line above

**Comment by ffont** [comment]
I'm waiting for @xavierfav to add some tests before I review and merge.

**Comment by ffont** [comment]
Is this now ready for re-review?

**Comment by xavierfav** [comment]
It is not. I struggle with "simulating" the post request. I do not get the right response.

**Comment by xavierfav** [comment]
Ready for review

File: datasets/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            form_data['form-{0}-visited_sound'.format(i)] = ['True', 'False'][i%2]
+            form_data['form-{0}-annotation_id'.format(i)] = str(all_annotation_object_ids[i])
+            form_data['form-{0}-vote'.format(i)] = ['1', '-1', '1'][i%3]
+            
+        # login
+        self.client.login(username='username_0', password='123456')
+        
+        # post requests
+        response = self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
+        response2 = self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
```
Comment (alastair) [inline datasets/tests/test_views.py]: If you're not going to use `response2`, you don't need to assign this.

File: datasets/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        # checking duplicates
+        duplicate_votes = []
+        for row in Vote.objects.all():
+            if Vote.objects.filter(created_by=row.created_by,
+                    annotation_id = row.annotation_id, 
+                    created_at__lte=row.created_at+timedelta(seconds=10),  
+                    created_at__gt=row.created_at-timedelta(seconds=10)).count() > 1:
+                duplicate_votes.append(row)
+                
+        self.assertEquals(response.status_code, 200)
```
Comment (alastair) [inline datasets/tests/test_views.py]: I'm looking at the view for this test, and see that there is nothing to return a different response code when there is an error.
We should change this so that when we return an error (https://github.com/MTG/freesound-datasets/blob/367d6f5702f0ed44ebc9b999dd3b5fb6f5a69929/datasets/views.py#L184) we also use the correct HTTP code (perhaps 400), and also add a test for this case.

File: datasets/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        
+        # create form data
+        form_data = {'form-MAX_NUM_FORMS': '1000', 
+                     'category_id': node_id, 
+                     'form-INITIAL_FORMS': nb_annotations, 
+                     'form-TOTAL_FORMS': nb_annotations + 1,
+                     'comment': '',
+                     'dataset': '1',
+                    }
+        for i in range(nb_annotations):
```
Comment (alastair) [inline datasets/tests/test_views.py]: In tests you should make the case that you are trying to check as simple as possible. In this case, we should only have 12 forms if this is a strict requirement which is enforced server-side in the view.
I believe this is not the case, so my preference is to have only one form in this formset.

File: datasets/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                     'form-TOTAL_FORMS': nb_annotations + 1,
+                     'comment': '',
+                     'dataset': '1',
+                    }
+        for i in range(nb_annotations):      
+            form_data['form-{0}-visited_sound'.format(i)] = ['True', 'False'][i%2]
+            form_data['form-{0}-annotation_id'.format(i)] = str(all_annotation_object_ids[i])
+            form_data['form-{0}-vote'.format(i)] = ['1', '-1', '1'][i%3]
+            
+        # login
```
Comment (alastair) [inline datasets/tests/test_views.py]: This is another comment that's not needed, because it doesn't tell us anything that we didn't already know

File: datasets/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        self.client.login(username='username_0', password='123456')
+        
+        # post requests
+        response = self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
+        response2 = self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
+
+        # checking duplicates
+        duplicate_votes = []
+        for row in Vote.objects.all():
```
Comment (alastair) [inline datasets/tests/test_views.py]: Again, in our test we know exactly what we are trying to submit twice. Our check in the view is based on the user and annotation_id. In this case, we can do two checks. One after the first submission (to see if we have 1 row for this user/annotationid) and another after the second submission (to check that we still have just 1)

File: datasets/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+from datasets.models import *
+from datasets.views import *
+from datasets.forms import *
+from datasets.management.commands.generate_fake_data import create_sounds, create_users, create_annotations
+from django.http.request import QueryDict, MultiValueDict
+from datetime import timedelta
+
+
+class ContributeTest(TestCase):
+    fixtures = ['datasets/fixtures/initial.json']
```
Comment (ffont) [inline datasets/tests/test_views.py]: However I must say that I'm happy with the full path being there, in my opinion it makes it clearer where the data comes from.

File: datasets/tests/test_views.py (hunk 3 above)
Comment (ffont) [inline datasets/tests/test_views.py]: If this is to be changed we should handle it also in the template code.
Right now the javascript only handles the `success` case (HTTP 200, https://github.com/MTG/freesound-datasets/blob/master/templates/contribute_validate_annotations_category.html#L12). 
If we add `status=400` when the forms are not valid and there are errors, then this should be handled in an `error` case in the javascript.

Comment (alastair) [inline datasets/tests/test_views.py]: Additionally we should decide what to do with the disabled submit button if the request fails... do we leave it disabled? Or do we re-enable it?

**Comment by xavierfav** [comment]
I did the following changes:
- Cleaning non useful comments.
- Add documentation to functions in generate_fake_data.py.
- Fixed the problem with the disable/unable submit button.
- Changed the strategy of test. I do three checks: (i) after submitting the 1st form, the response is the one expected, (ii) the vote has been added to the database, (iii) after a second submission, no vote is added in the database.
- Simplified the test. One fake sound and annotation created only, one annotation vote form.
- Changed print() to logging.info() (I have not done any configuration of the logging, I don't really know what is the best here and what should be exactly done)

File: datasets/tests/test_views.py [hunk 8]
```diff
@@ -3,56 +3,44 @@
...
-                
+        # check the response an that a vote is added in the database
+        response = self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data) 
         self.assertEquals(response.status_code, 200)
-        self.assertEquals(duplicate_votes, [])
+        self.assertEquals(len(Vote.objects.filter(annotation_id=annotation_object_id)), 1)
+        
+        # check that a second vote is not added
+        self.client.post(reverse('save-contribute-validate-annotations-per-category'), data=form_data)
+        self.assertEquals(len(Vote.objects.filter(annotation_id=annotation_object_id)), 1)
```
Comment (ffont) [inline datasets/tests/test_views.py]: Maybe here you should check also if the value of the vote has changed. Eg, send vote 1 with vote value "1", send vote 2 with vote value "-1", check that stored vote is still "1".

Besides that all the rest look great to me. If you add this little fix I'll merge and deploy :)

File: datasets/management/commands/generate_fake_data.py [hunk 9]
```diff
@@ -4,16 +4,35 @@
...
 
 def create_sounds(dataset_short_name, num_sounds):
-    # Get the dataset
+    """Create fake Sound object instances.
+    
+    Create fake sounds to the dataset instance with the specified short name.
+    
+    Args:
+        dataset_short_name: A string corresponding to the short name of the dataset to which the sounds will be added.
+        num_sounds: An integer corresponding to the number of fake sounds to create.
```
Comment (alastair) [inline datasets/management/commands/generate_fake_data.py]: I don't think these comments need to be so verbose - 
```
dataset_short_name: The short name of the dataset to which the sounds will be added.
num_sounds: The number of fake sounds to create.
```

but don't worry about changing them, just keep in mind for the future.

File: templates/contribute_validate_annotations_category.html [hunk 10]
```diff
@@ -5,7 +5,7 @@
 {% block page_js %}
     <script type="text/javascript">
         function submitForm(){
-            $("#submitButton").prop('disable', true);
+            document.getElementById("submitButton").disabled = true;
```
Comment (alastair) [inline templates/contribute_validate_annotations_category.html]: out of interest, why did you change this from jquery (`$("....")`) to plain javascript (`document.getElementById`)? Both are correct, but we're using jquery in other places here anyway, so its use here is not a problem

File: datasets/tests/test_views.py [hunk 11]
```diff
@@ -11,36 +11,34 @@ class ContributeTest(TestCase):
...
+                     'form-TOTAL_FORMS': 2,
                      'comment': '',
                      'dataset': '1',
-                    }
-        for i in range(nb_annotations):      
-            form_data['form-{0}-visited_sound'.format(i)] = ['True', 'False'][i%2]
-            form_data['form-{0}-annotation_id'.format(i)] = str(all_annotation_object_ids[i])
-            form_data['form-{0}-vote'.format(i)] = ['1', '-1', '1'][i%3]
+                    }     
+        form_data['form-0-visited_sound'] = 'Flase'
```
Comment (alastair) [inline datasets/tests/test_views.py]: False

What does this field in the form mean? Is it added to the database anywhere?

File: datasets/tests/test_views.py (hunk 8 above)
Comment (alastair) [inline datasets/tests/test_views.py]: Yes, agreed.

Django also has a `.count()` method which you can call on a Query object, which will give you the same number without having to call `len`

**Comment by alastair** [comment]
Please look at your configured name and email in git: https://help.github.com/articles/setting-your-username-in-git/

File: templates/contribute_validate_annotations_category.html (hunk 10 above)
Comment (xavierfav) [inline templates/contribute_validate_annotations_category.html]: Using the prop() jquery method did not work.
I switched to the native way of doing it

**Comment by xavierfav** [comment]
Changes done
